File: week10/19/P1.py
class Solution(object):
    def getMoneyAmount(self, n):
        memo = {} # 중복은 제외하기 위한 메모이제이션

        def dp(left, right):
            if (left, right) in memo:
                return memo[(left, right)]
            if left >= right:
                return 0
            ans = float('inf')
            for pick in range(left, right + 1):
                leftCost = dp(left, pick - 1) + pick
                rightCost = dp(pick + 1, right) + pick
                cost = max(leftCost, rightCost)
                ans = min(ans, cost)
            memo[(left, right)] = ans
            return ans

        return dp(1, n)

if __name__ == "__main__":
    result = Solution()
    print(result.getMoneyAmount(56))
    # print(result.getMoneyAmount(16))
    # print(result.getMoneyAmount(10))
    # print(result.getMoneyAmount(5))
    # print(result.getMoneyAmount(2))
    # print(result.getMoneyAmount(1))

File: week10/19/P1.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution {
public:
    unordered_map<int, int> memo;

    int dp(int left, int right) {
        int key = left * 1000 + right; // left, right를 하나의 고유 키로 만들기 위해
        if (memo.find(key) != memo.end()) return memo[key];
        if (left >= right) return 0;
        int ans = INT_MAX;
        for (int pick=left; pick<right+1; pick++) {
            int leftCost = dp(left, pick - 1) + pick;
            int rightCost = dp(pick + 1, right) + pick;
            int cost = max(leftCost, rightCost);
            ans = min(ans, cost);
        }
        memo[key] = ans;
        return ans;
    }

    int getMoneyAmount(int n) {
        return dp(1, n);
    }
};

int main() {
    Solution result;
    cout << result.getMoneyAmount(56) << endl;

    return 0;
}